### PR TITLE
feat: Deploy self-monitor via telemetry manager

### PR DESCRIPTION
**Pull request by rakesh-garimella**
## Description

Changes proposed in this pull request (what was done and why):

- Deploy Selfmonitor via code
-  Only deploy when we have pipeline from any telemetry signal. Remove the self-monitor when there is no pipeline
- Selfmonitor is deployed for logpipelines/metricpipeline/tracepipelines
- Deploy following resources:
    - Self-monitor deployment/ configmap
    - Role/Role Binding/Service Account
    - Network Policy
- Use dummy config for now
- Some refactoring

Changes refer to particular issues, PRs or documents:

- https://github.com/kyma-project/telemetry-manager/issues/807

## Traceability
- [x] The PR is linked to a GitHub issue.
- [ ] New features have a milestone set.
- [ ] New features have defined acceptance criteria in a corresponding GitHub Issue, and all criteria are satisfied with this PR.
- [x] The corresponding GitHub issue has a respective `area` and `kind` label.
- [ ] The follow-up issues (if any) are linked in the `Related Issues` section.
- [ ] Adjusted the documentation if the change is user-facing.
- [x] The feature is unit-tested
- [ ] The feature is e2e-tested

### CODE REVIEW

File: internal/selfmonitor/config.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+type monitoringConfig struct {
+	GlobalConfig   prometheus.GlobalConfig   `yaml:"global"`
+	AlertingConfig prometheus.AlertingConfig `yaml:"alerting,omitempty"`
+	RuleFiles      []string                  `yaml:"rule_files,omitempty"`
+	ScrapeConfigs  []prometheus.ScrapeConfig `yaml:"scrape_configs,omitempty"`
+}
+
+func (promCfg *Config) WithMonitoringConfig(monitoringCfgYAML string) *Config {
```
Comment (chrkl) [inline internal/selfmonitor/config.go:33]: I don't understand why we need to return a copy. Also the name indicates the use of the functional options pattern, which is not given here. Maybe see if we can just drop this and make a simple assignment where it is used.

File: internal/reconciler/logpipeline/reconciler.go [hunk 2]
```diff
@@ -222,6 +233,25 @@ func (r *Reconciler) reconcileFluentBit(ctx context.Context, pipeline *telemetry
 	return nil
 }
 
+func (r *Reconciler) reconcileSelfMonitor(ctx context.Context, pipeline *telemetryv1alpha1.LogPipeline, deployableLogPipelines []telemetryv1alpha1.LogPipeline) error {
```
Comment (chrkl) [inline internal/reconciler/logpipeline/reconciler.go:236]: Maybe consider moving this to the telemetry controller. Managing one resource by multiple controllers has always the side-effect that one reconciliation will trigger the other controllers and can easily lead to an infinite reconciliation loop. Also that would remove some code redundancy.

File: internal/selfmonitor/deployment.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/telemetry-manager/internal/configchecksum"
+	"github.com/kyma-project/telemetry-manager/internal/k8sutils"
+)
+
+const (
+	prometheusUser         int64 = 10001
+	collectorContainerName       = "self-monitor"
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply internal/selfmonitor/deployment.go:25]: done

File: internal/reconciler/metricpipeline/reconciler.go [hunk 4]
```diff
@@ -48,16 +50,18 @@ type Reconciler struct {
 	agentProber        DaemonSetProber
 	overridesHandler   *overrides.Handler
 	istioStatusChecker istiostatus.Checker
+	enableSelfMonitor  bool
 }
 
-func NewReconciler(client client.Client, config Config, gatewayProber DeploymentProber, agentProber DaemonSetProber, overridesHandler *overrides.Handler) *Reconciler {
+func NewReconciler(client client.Client, config Config, gatewayProber DeploymentProber, agentProber DaemonSetProber, overridesHandler *overrides.Handler, enableSelfMonitor bool) *Reconciler {
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply internal/reconciler/metricpipeline/reconciler.go:56]: done

File: internal/selfmonitor/config_builder.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+func MakeConfig() monitoringConfig {
+	promConfig := monitoringConfig{}
+	promConfig.GlobalConfig = globalConfigBuilder()
+	promConfig.AlertingConfig = alertConfigBuilder()
+	promConfig.RuleFiles = []string{"/etc/prometheus/prometheus.rules"}
+	promConfig.ScrapeConfigs = scrapeConfigBuilder()
+	return promConfig
+}
+
+func globalConfigBuilder() prometheus.GlobalConfig {
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply internal/selfmonitor/config_builder.go]: done

File: internal/prometheus/relabel_configs.go [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package agent
+package prometheus
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply internal/prometheus/relabel_configs.go]: done

File: internal/selfmonitor/deployment.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/telemetry-manager/internal/configchecksum"
+	"github.com/kyma-project/telemetry-manager/internal/k8sutils"
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply internal/selfmonitor/deployment.go]: done

File: internal/reconciler/logpipeline/reconciler.go (hunk 2 above)
Comment (rakesh-garimella) [reply internal/reconciler/logpipeline/reconciler.go]: done

File: internal/selfmonitor/deployment.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
+package selfmonitor
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply internal/selfmonitor/deployment.go]: done

**Review by chrkl — APPROVED** [review]

**Pull request merged**